### DISCUSSION TITLE
making GUI tool and Windows and macos support

### DISCUSSION THREAD

**Discussion by Vaisakhkm2625** [xremap/xremap#358] [Q&A] [answered]
I have been planning to make a keyboard remapping and automation tool like "BetterTouchTool" on macos, but for linux and windows, as a university project for one of my friend... 
mean while  i came across 22+ keyboard remapping tools, along with xremap (found out about xremap from [vimjoyer's video](https://youtu.be/lyxScRCe6bE) ) and really liked this...

So we thinking of making a PyQT based ui for this (as he is not familiar with rust and if need could be ported to rust later by using it as a first iteration ). Along with implementing support for windows and macos if possible. 

My question is, is there any support for window or mac already there in xremap? 
Is it a stupid idea to make a gui tool like this?

**Comment by k0kubun** [comment, marked as answer]
> found out about xremap from [vimjoyer's video](https://youtu.be/lyxScRCe6bE)

😮 I didn't know there's a video about xremap. Thank you for letting me know.

> My question is, is there any support for window or mac already there in xremap?

No, it's solely relying on evdev, which is specific to Linux (and FreeBSD). It's already too hard for a single maintainer to support various Linux platforms, so I wouldn't support other OSes unless somebody steps up to own that portion.

> Is it a stupid idea to make a gui tool like this?

Not at all 🙂 It's kinda designed with that in mind from the beginning. All it needs to do would be to generate a YAML (or JSON) file, so it's probably not difficult. You can use `--watch` (or `--watch=config`) to integrate the GUI app with xremap. So it doesn't need to be Rust or in the same repository.

> So we thinking of making a PyQT based ui for this (as he is not familiar with rust and if need could be ported to rust later by using it as a first iteration ). Along with implementing support for windows and macos if possible.

If I were to implement it myself, I would use Electron since [I've already used that](https://github.com/k0kubun/Nocturn) and I'm more familiar with HTML/CSS/JavaScript than QT API or any other GUI kits. But feel free to choose whatever you want for your one.

**Reply by Vaisakhkm2625** [reply]
😅 thanks for your reply...

> Not at all 🙂 It's kinda designed with that in mind from the beginning

yes i saw examples of bindings to other languages... that's also reason why thought of using this fantastic tool
> so I wouldn't support other OSes unless somebody steps up to own that portion.

😃 we could try that... or at-least make a prototype for that...
 >  I would use Electron

In retrospect, that would have been a better idea, and would have been much easier... but my friend only knows basic c,java and python... :( and it's his project... i actually wanted to go with Tauri and Yew or iced... but anyway i wanted to learn qt because i wanted to contribute to KDE...

> since [I've already used that](https://github.com/k0kubun/Nocturn) 

wow.. that really cool app... seems really polished...
